common/environment/environment.cpp

```diff
@@ -2615,6 +2614,39 @@ void getRoxieProcessServers(const char *process, SocketEndpointArray &servers)
     getRoxieProcessServers(queryRoxieProcessTree(root, process), servers);
 }
 
+extern ENVIRONMENT_API RoxieTargetType readRoxieTargetType(const char *roxieName)
+{
+    RoxieTargetType roxieTargetType = RTTUnknown;
+    if (isEmptyString(roxieName))
+        return roxieTargetType;
+
+    bool hasPortZero = false, hasNonZeroPort = false;
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
+    Owned<IConstEnvironment> environment = factory->openEnvironment();
+    Owned<IPropertyTree> envRoot = &environment->getPTree();
+
+    VStringBuffer xpath("Software/RoxieCluster[@name='%s'][1]", roxieName);
+    Owned<IPropertyTreeIterator> farmers = envRoot->queryPropTree(xpath.str())->getElements("RoxieFarmProcess");
+    ForEach(*farmers)
+    {
+        IPropertyTree &farmer = farmers->query();
+        const char *port = farmer.queryProp("@port");
+        if (!port)
+            continue;
+        if (streq(port, "0"))
+            hasPortZero = true;
+        else
+            hasNonZeroPort = true;
+    }
```

Comment:
You could stop the loop once you have both.

Reply:
Will do